include/zephyr/sys/util.h

```diff
@@ -26,6 +26,88 @@
+	} else if (sizeof(x) <= sizeof(unsigned long)) {
+		return (x > 1) * NUM_BITS(unsigned long) - __builtin_clzl(x - 1);
+	} else {
+		return (x > 1) * NUM_BITS(unsigned long long) - __builtin_clzll(x - 1);
+	}
+
+	__ASSERT_NO_MSG(sizeof(x) == sizeof(unsigned long long));
+}
+
+template <typename T> static inline constexpr uint64_t __z_nhpot_impl(T x)
+{
+	int l2c = __z_log2ceil_impl(x);
+
+	return (x == 0) + (x != 0 && l2c != 64) * (1ULL << l2c);
+}
+#else
+#define __z_log2_impl(x)                                                                           \
+	(((x) < 1) * (-1) +                                                                        \
+	 ((x) >= 1) * _Generic((x), char                                                           \
+			       : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), unsigned char    \
+			       : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), short            \
+			       : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), unsigned short   \
+			       : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), int              \
+			       : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), unsigned int     \
+			       : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), long             \
+			       : (NUM_BITS(unsigned long) - __builtin_clzl(x) - 1), unsigned long  \
+			       : (NUM_BITS(unsigned long) - __builtin_clzl(x) - 1), long long      \
+			       : (NUM_BITS(unsigned long long) - __builtin_clzll(x) - 1),          \
+				 unsigned long long                                                \
+			       : (NUM_BITS(unsigned long long) - __builtin_clzll(x) - 1)))
+
+#define __z_log2ceil_impl(x)                                                                       \
+	(((x) > 1) *                                                                               \
+	 _Generic((x), char                                                                        \
+		  : (NUM_BITS(unsigned int) - __builtin_clz((x)-1)), unsigned char                 \
+		  : (NUM_BITS(unsigned int) - __builtin_clz((x)-1)), short                         \
+		  : (NUM_BITS(unsigned int) - __builtin_clz((x)-1)), unsigned short                \
+		  : (NUM_BITS(unsigned int) - __builtin_clz((x)-1)), int                           \
+		  : (NUM_BITS(unsigned int) - __builtin_clz((x)-1)), unsigned int                  \
+		  : (NUM_BITS(unsigned int) - __builtin_clz((x)-1)), long                          \
+		  : (NUM_BITS(unsigned long) - __builtin_clzl((x)-1)), unsigned long               \
+		  : (NUM_BITS(unsigned long) - __builtin_clzl((x)-1)), long long                   \
+		  : (NUM_BITS(unsigned long long) - __builtin_clzll((x)-1)), unsigned long long    \
+		  : (NUM_BITS(unsigned long long) - __builtin_clzll((x)-1))))
+
+#define __z_nhpot_impl(x)                                                                          \
+	(((x) == 0) + ((x) != 0) * (_Generic((x), char                                             \
```

Comment:
I might have it wrong, but I think the `((x) == 0) + ((x) != 0) *` part is not needed. The case of x == 0 is handled correctly with `1 << LOG2CEIL(x)` if the intention is that it returns `1`.

Reply:
Yep - you were right @oyvindronningstad - thanks for catching that 😁

Reply:
Did you mean to change this? (It's still the same)

Reply:
@cfriedt Still curious about this ^.

Otherwise LGTM.

Reply:
I'll double check. It's possible that I might have forgotten to synchronize on a different machine and undid it by accident.

Reply:
@oyvindronningstad - fixed!